160 remove complex widget

MPParsley opened the pull request:
Move amitaibu:160-remove-complex-widget & damienmckenna:160-remove-complex-widget to Gizra:og

See #181 & #558.

Fixes #160

amitaibu committed:
Start removing complex widget

amitaibu committed:
Add core widgets

amitaibu committed:
Allow privileged user to create content

amitaibu committed:
Check permissions

amitaibu committed:
Start fixing tests

amitaibu committed:
Fix logic and tests

amitaibu committed:
Sniffer fixes

amitaibu committed:
Fix return value in docs

amitaibu committed:
Re-add Functional test

amitaibu committed:
Sniffer fixes

amitaibu committed:
Sniffer fixes

amitaibu committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Setting up the base for the new test.

Commit:
Commit more work.

Commit:
First version of the tests.

Commit:
Make code sniffer happy.

amitaibu committed:
Merge 8.x-1.x

amitaibu committed:
Remove un-needed method

amitaibu committed:
Remove unused statement

amitaibu committed:
Start expanding the widget tests

amitaibu committed:
Create node types

amitaibu committed:
Comments cleanup

amitaibu committed:
Remove traits

amitaibu committed:
Try to find cause of bug

amitaibu committed:
Remove unusued code

amitaibu committed:
Remove unusued code

amitaibu committed:
Remove custom settings

amitaibu committed:
Remove custom settings

amitaibu committed:
Remove unused use

amitaibu committed:
Add todo

amitaibu committed:
Filter out the bundles that are not groups.

amitaibu committed:
Filter out the bundles that are not groups.

amitaibu committed:
Fix indentation

amitaibu committed:
Fix indentation

amitaibu committed:
Improve exception error

amitaibu committed:
Improve names

amitaibu committed:
Sniffer fixes

amitaibu committed:
Convert test to Functional

amitaibu committed:
Add comment

amitaibu committed:
Debug failing tests

amitaibu committed:
Revert debug

amitaibu committed:
Fix wrong node type name

amitaibu committed:
Simplify and debug test

amitaibu committed:
Add status 200

amitaibu committed:
gst

amitaibu committed:
Quicker debugs

amitaibu committed:
Check 404 page

amitaibu committed:
Rebuild routes

amitaibu committed:
Fields page

amitaibu committed:
click link

amitaibu committed:
Back to drupalGet

amitaibu committed:
Check access

amitaibu committed:
Enable field ui

amitaibu committed:
Allow debug

amitaibu committed:
Import URL

amitaibu committed:
Debug user access

amitaibu committed:
Cleanup

amitaibu committed:
Admin user

amitaibu committed:
Back to permissions

amitaibu committed:
Remove use

amitaibu committed:
Permissions

amitaibu committed:
Core test

amitaibu committed:
Run core

amitaibu committed:
Debug core

amitaibu committed:
Remove core debug

amitaibu committed:
Start conveting test to Kernel

amitaibu committed:
More work

amitaibu committed:
Use entity test with bundle

amitaibu committed:
Merge field settings

amitaibu committed:
Fix test

amitaibu committed:
Sniffer fixes

amitaibu committed:
Remove debug

amitaibu committed:
Create Node type correctly

amitaibu committed:
Add unpublished node

amitaibu committed:
Fix OgSelectionWidgetTest

Commit:
Validate that the user have access to post a group content.

Commit:
Adding skeleton for a test [skip ci]

Commit:
Commit work [skip ci]

Commit:
Fix a typo and continue working on the test [skip ci]

Commit:
Set the role properly [skip ci]

Commit:
Use forbidden and not allowed.

Commit:
Checkin when forbidden.

Commit:
Somehow fix the tests.

Commit:
Getting the root entity.

Commit:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Fix conflict

Commit:
Stbalize tests.

Commit:
Minor refactor.

Commit:
Wrong operator.

Commit:
Trying to understand what’s the difference [skip ci]

Commit:
Find another solution.

Commit:
Adding basic constraint to add entities [skip ci]

Commit:
Commit logic for now.

Commit:
Don’t add constant to all the entities, just check the field.

Commit:
Fix tests.

Commit:
Remove the testin file

Commit:
Fix name.

Commit:
Fix cs

Commit:
Get the text.

Commit:
Maybe now?

Commit:
Catching and not version.

Commit:
CS fix.

Commit:
Handle another case for the fields.

Commit:
Set the field as required in specific cases.

Commit:
More changes.

Commit:
Adding the validation.

Commit:
The validation works as expected. Now, we need to write tests.

Commit:
More improvements.

Commit:
Adding the boilerplate for the new test.

Commit:
Done with the tests.

Commit:
Minor fixes.

Commit:
Make CS happy.

Commit:
make cs happy.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 160-remove-complex-widget

mariano-dagostino committed:
Fix grammar in comments.

mariano-dagostino committed:
Improve comment of access check logic

mariano-dagostino committed:
GroupTypeManager::getAllGroupBundles should respect GroupTypeManagerInterface method signatures.

RoySegall committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Use single quotes

MPParsley committed:
Undo single quotes

MPParsley committed:
Improve readability foreach

MPParsley committed:
Updated documentation for the entity reference field

MPParsley committed:
Use invalidMessage from constraint

MPParsley committed:
Added dependency injection to ValidOgMembershipMultipleReferenceConstraintValidator

pfrenssen committed:
Merge remote-tracking branch 'gizra/8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Fix tests

MPParsley committed:
Fixed unnecessarily gendered language in a comment

MPParsley committed:
Cleanup call to MembershipManager::reset()

Removed in #480.

MPParsley committed:
Protected properties should start with a small cap

MPParsley committed:
Protected properties should start with a small cap.

MPParsley committed:
Update src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraintValidator.php

MPParsley committed:
Revert commit

MPParsley committed:
Deprecate Drupal\simpletest\ContentTypeCreationTrait

See https://www.drupal.org/node/2884454

MPParsley committed:
Deprecated Drupal\simpletest\NodeCreationTrait

See https://www.drupal.org/node/2884454

MPParsley committed:
Fixed namespaces

MPParsley committed:
Removed deprecated call to Drupal::entityManager

MPParsley committed:
Removed deprecated entity.query service

MPParsley committed:
Removed calls to deprecated entityManager

MPParsley committed:
Use EntityTypeManagerInterface

MPParsley committed:
Remove call to deprecated entity.manager

MPParsley committed:
Fixed EntityManager deprecations in OgSelection

See https://www.drupal.org/node/2549139

MPParsley committed:
Fixed call to undefined function getBundleInfo() on EntityTypeManager

MPParsley committed:
Added getQuery

MPParsley committed:
Clear membership cache

MPParsley committed:
Restore cache reset

MPParsley committed:
Test with cache delete.

MPParsley committed:
Cleanup test

MPParsley committed:
Fixed indentation

MPParsley committed:
Added missing use statement

MPParsley committed:
Fix tests

MPParsley committed:
Revert permission format

MPParsley committed:
Revert permission format

MPParsley committed:
Revert permission format

damienmckenna committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Merge pull request #569 from damienmckenna/160-remove-complex-widget

160 remove complex widget

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -2,8 +2,19 @@
+use Drupal\Core\Session\AccountInterface;
+use Drupal\og\MembershipManagerInterface;
+use Drupal\og\OgAccessInterface;
+use Drupal\user\Entity\User;
```
Remove unused use statement:
```suggestion
```

MPParsley committed:
Remove unused use statement

pfrenssen commented on src/Plugin/Field/FieldType/OgStandardReferenceItem.php:
```diff
@@ -3,50 +3,22 @@
-      '#description' => $this->t('By default, the <em>administer group</em> permission is required to directly edit this field. Selecting this option will allow access to anybody with access to edit the entity.'),
-      '#type' => 'checkbox',
-      '#default_value' => $this->getSetting('access_override'),
-    ];
```
I verified this and this setting is not used anywhere and is OK to remove. However this is not in scope of this PR and it would make it easier to review this if we can move this to a separate PR.

MPParsley replied:
Fixed in #618.

pfrenssen commented on og.module:
```diff
@@ -159,6 +160,21 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
+      // Check if the field needs to be required.
+      $field_definition->setRequired($access);
+    }
+  }
```
What is the purpose of this change? It is not fully clear to me since this is not documented.

It looks like we are doing some checks and then set a field to be required. This is probably not the right place to do this. Since this is a `hook_entity_access()` implementation we should only have code here that is necessary for determining the `AccessResult` object to return. We should not mess with field definitions here.

MPParsley replied:
@amitaibu, can you confirm this or provide some more information where this comes from?

pfrenssen commented on og.module:
```diff
@@ -179,6 +195,11 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
 function og_entity_create_access(AccountInterface $account, array $context, $bundle) {
   $entity_type_id = $context['entity_type_id'];
 
+  if (!empty($context['skip_og_permission'])) {
+    // We've already been here or we want to skip the access check.
+    return;
+  }
```
Why is this necessary? This seems brittle to me and a potential security problem. We should not allow to skip access checks unless there is a very good reason for it.

amitaibu replied:
Not sure about the correctness of the code here - but I remember the reason though. When you create a node, it still doesn't belong to a group. It _might_ belong, but if for example the audience field is optional, than that node could be a "global" one (i.e. does not belong to a group).

So we need to check if that node could be created outside of OG context. If it can, then we don't have the final say.

pfrenssen replied:
It looks like this is intended to avoid an endless loop, because this is calling itself again later on.

This approach feels very strange to me. The code below calls the access check, passing exactly the same parameters, from inside the access check? What information could this possibly be retrieving that we don't already have at this point.

pfrenssen commented on og.module:
```diff
@@ -203,17 +224,32 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
     }
 
-    $handler = Og::getSelectionHandler($field_definition);
+    if (count($audience_fields) === 1) {
+      // Set the field to required only if we have a single audience field.
+      // If we have multiple fields then the require logic will perform at the
+      // entity constraint level.
+      $field_definition->setRequired(!\Drupal::entityTypeManager()
+        ->getAccessControlHandler($entity_type_id)
+        ->createAccess($bundle, $account, ['skip_og_permission' => TRUE]));
+    }
```
Why are we again setting fields to be required inside an access hook? This doesn't belong here.

amitaibu replied:
I believe it's related to https://github.com/Gizra/og/pull/570/files#r377001979

pfrenssen replied:
We should investigate a different solution, this is a hack. We should not alter any field definitions in `hook_entity_access()`.

It looks like this field definition is used as temporary storage for what could be a simple `$is_required` flag.

pfrenssen commented on src/GroupTypeManager.php:
```diff
@@ -191,8 +191,22 @@ public function getGroupBundleIdsByEntityType($entity_type_id) {
   public function getAllGroupBundles($entity_type = NULL) {
+    // Todo - should be remove since this method don't do any thing.
+    return $this->getGroupMap();
+  }
```
This change is not correct, it is losing the functionality to filter the group bundles because it is ignoring the `$entity_type` parameter.

It is out of scope for this PR to change existing functionality. Let's only make the changes here that are needed to remove the complex widget.

MPParsley replied:
Also see https://github.com/Gizra/og/pull/570/files#r339988596

MPParsley replied:
Reverted this change.

pfrenssen commented on src/GroupTypeManager.php:
```diff
@@ -191,8 +191,22 @@ public function getGroupBundleIdsByEntityType($entity_type_id) {
+   * @return array
+   *   An associative array of bundle IDs, or an empty array if none found.
+   */
+  public function getGroupBundlesByEntityType($entity_type_id) {
```
There is already a method for this: `getAllGroupBundles()`. I agree that this is a better name, but this is out of scope for this ticket. If we want to make this change let's do it in a dedicated PR and keep the focus of this PR on removing the complex widget.

MPParsley replied:
Also see https://github.com/Gizra/og/pull/181/files#r199885391

> Let's remove this method, it exactly duplicates the code from GroupTypeManager::getGroupsForEntityType().

MPParsley replied:
Reverted this change and fixed getAllGroupBundles in #402.

pfrenssen reviewed (changes requested):
I went through 1/4th of the code but it seems like this is introducing quite a lot of unrelated changes. This makes it quite difficult to review.

Also this is not only removing the complex widget but also introducing a new one. I think this can be split into two separate PRs, and since this is a big break from how OG used to work, the PR should including the documentation for end users on how to use the field.

For now, if we ever want to get this in can we please rework this so that it only contain the code that is strictly necessary to remove the complex widget?

It seems that some of unrelated changes are still useful though so we can probably move them to separate PRs.

damienmckenna commented:
Looking at the merge conflicts, the changes to config/schema/og.schema.yml and src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php are straight forward, I'm not sure about the other though.

damienmckenna committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

damienmckenna commented:
Please see #595 for my attempt to fix the merge conflicts on this branch.

pfrenssen committed:
Merge pull request #595 from damienmckenna/feature/160-remove-complex-widget

Fix merge conflicts from 8.x-1.x

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraint.php:
```diff
@@ -0,0 +1,28 @@
+   *
+   * @var string
+   */
+  public $AudienceFieldsAreNotPopulated = 'One of the fields @fields is required.';
```
Class property $AudienceFieldsAreNotPopulated should use lowerCamel naming without underscores

```suggestion
  public $audienceFieldsAreNotPopulated = 'One of the fields @fields is required.';
```

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraint.php:
```diff
@@ -23,4 +23,11 @@ class ValidOgMembershipReferenceConstraint extends Constraint {
+   *
+   * @var string
+   */
+  public $NotAllowedToPostInGroup = 'You are not allowed to post content in the group %label';
```
Class property $NotAllowedToPostInGroup should use lowerCamel naming without underscores

```suggestion
  public $notAllowedToPostInGroup = 'You are not allowed to post content in the group %label';
```

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +20,34 @@ public function validate($value, Constraint $constraint) {
+    $access = \Drupal::service('og.access')->userAccessEntity($permission, $group, $user);
+
+    if ($access->isForbidden()) {
+      $this->context->addViolation($constraint->NotAllowedToPostInGroup, $params);
```
Class property $NotAllowedToPostInGroup should use lowerCamel naming without underscores

```suggestion
      $this->context->addViolation($constraint->notAllowedToPostInGroup, $params);
```

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraintValidator.php:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    if ($fields_are_empty) {
+      $this->context->addViolation('One of the fields @fields is required.', [
```
Use the constraints error message:

```suggestion
      $this->context->addViolation($constraint->audienceFieldsAreNotPopulated, [
```

MPParsley committed:
Fix codesniffer errors

use lowerCamel naming for class properties

MPParsley commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,181 @@
+    $this->drupalGet('node/add/group_content');
+    // The user can't post group content site wide so the audience is required
+    // field.
+    // todo: is that correct?
```
See https://github.com/Gizra/og/pull/181/files#r112639472

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,46 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
-    if (!empty($this->getConfiguration()['field_mode']) && $this->getConfiguration()['field_mode'] === 'admin') {
-      // Don't include the groups, the user doesn't have create permission.
-      foreach ($user_groups as $delta => $group) {
+    foreach ($this->getUserGroups() as $group) {
```
See https://github.com/Gizra/og/pull/181/files#r116233873

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,46 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
 
-    $identifier_key = $definition->getKey('id');
+    if (empty($this->configuration['entity'])) {
+      // @todo: Find out why we have this scenario.
```
See https://github.com/Gizra/og/pull/181/files#r111669853

damienmckenna committed:
Removed redundant getGroupBundlesByEntityType() method.

damienmckenna committed:
Typoo.

MPParsley committed:
Merge pull request #596 from damienmckenna/feature/570-getGroupBundlesByEntityType

Removed redundant getGroupBundlesByEntityType() method.

MPParsley committed:
Revert "Removed redundant getGroupBundlesByEntityType() method."

MPParsley committed:
Merge pull request #597 from Gizra/revert-596-feature/570-getGroupBundlesByEntityType

Revert "Removed redundant getGroupBundlesByEntityType() method."

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Fix merge conflict

MPParsley commented on og.module:
```diff
@@ -154,6 +155,21 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
+  $audience_fields = \Drupal::service('og.group_audience_helper')->getAllGroupAudienceFields($entity_type_id, $bundle_id);
+  if (count($audience_fields) === 1) {
+    $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity_type_id, $bundle_id);
+    foreach ($field_definitions as $field_name => $field_definition) {
```
Unused variable $field_name.
```suggestion
    foreach ($field_definitions as $field_definition) {
```

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,46 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
+
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    $entity = $this->configuration['entity'];
+    $entity_type_id = $entity->getEntityTypeId();
```
Unused variable $entity_type_id.
```suggestion
```

MPParsley committed:
Removed unused variables

MPParsley committed:
Update og.module

MPParsley commented:
The [merge](https://github.com/Gizra/og/pull/570/commits/8a35d7a1cc4e276af0a40298dc9d30b760dec193) of #610 has broken some tests here:
- OgSelectionWidgetOptionsTest::testNonRequiredAudienceField (OgSelectionWidgetOptionsTest.php:144)
- OgSelectionTest::testSelectionHandlerResults (OgSelectionTest.php:196)

@damienmckenna, @kbrodej, @pfrenssen can you have a look?

edit: after some debugging it seems that this PR is altering the deprecated Og::getSelectionHandler. Not sure what's the best way to proceed here.

MPParsley commented on og.module:
```diff
@@ -198,23 +219,39 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
     $options = [
       'target_type' => $field_definition->getFieldStorageDefinition()->getSetting('target_type'),
       'handler' => $field_definition->getSetting('handler'),
       'field_mode' => 'admin',
```
```suggestion
```

See https://github.com/Gizra/og/pull/610/files#r376804037

MPParsley committed:
Test without admin field_mode

MPParsley commented on og.module:
```diff
@@ -198,23 +219,38 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
 
-    if ($handler->createInstance('og:default', $options)) {
-      return AccessResult::neutral();
+    if ($handler && $handler->countReferenceableEntities()) {
```
```suggestion
    // @todo: remove deprecated call to Og::getSelectionHandler.
    $handler = Og::getSelectionHandler($field_definition);
    if ($handler && $handler->countReferenceableEntities()) {
```

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -123,12 +134,29 @@ protected function setUp() {
     $this->selectionPluginManager->createInstance('og:default', $options);
     $this->selectionHandler = $this->selectionPluginManager->getSelectionHandler($this->fieldDefinition);
```
```suggestion
    $this->selectionHandler = $this->selectionPluginManager->getSelectionHandler($this->fieldDefinition);
    // @todo: remove deprecated call to Og::getSelectionHandler.
    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```

MPParsley committed:
Test with deprecated Og::getSelectionHandler

MPParsley committed:
fix merge conflict

amitaibu commented on og.module:
```diff
@@ -154,6 +155,21 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
+      }
+
+      // Check if the field needs to be required.
+      $field_definition->setRequired($access);
```
@MPParsley this one is related to https://github.com/Gizra/og/pull/570/files#r377001979

We try to make sure that if a node must be created inside an OG context, it cannot be created without choosing a group. Again not sure about the correctness of code - but rather on the reason.

pfrenssen replied:
I think this code is not working. This is calling `FieldConfigInterface::setRequired()` but never calling `FieldConfigInterface::save()` after making the change so this code is basically doing nothing.

We should not alter field definitions in `hook_entity_access()` in any case.

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

pfrenssen commented:
I have tried reviewing this already several times now, but always when I get through the first sections I end up with a bad feeling. There are some workarounds being introduced in the access checking that are potentially insecure, probably not working as intended, and possibly having a performance impact.

I think maybe the best approach would be to revert the code changes in the access hooks, check that our test coverage is on point (i.e. before we make changes to access handling we should first have some failing tests that show which access checks are missing), and then reimplement the needed access checks to make the tests green, preferably without too many strange code acrobatics :)

MPParsley commented:
> we have test failing due to missing access checks

@pfrenssen, as it now stands the reason tests are failing is because we're using some deprecations, no?

```
Testing /home/travis/build/Gizra/drupal/modules/og/tests
...............................................................  63 / 565 ( 11%)
............................................................... 126 / 565 ( 22%)
............................................................... 189 / 565 ( 33%)
............................................................... 252 / 565 ( 44%)
............................................................... 315 / 565 ( 55%)
............................................................... 378 / 565 ( 66%)
............................................................... 441 / 565 ( 78%)
............................................................... 504 / 565 ( 89%)
.............................................................   565 / 565 (100%)
Time: 3.39 minutes, Memory: 110.00MB
OK (565 tests, 4789 assertions)
Remaining deprecation notices (48)
  22x: Og:getSelectionHandler() is deprecated in og:8.x-1.0-alpha4
      and is removed from og:8.x-1.0-alpha5.
      Use \Drupal\Core\Entity\EntityReferenceSelection\SelectionPluginManager::getInstance()
      instead. See https://github.com/Gizra/og/issues/580
    10x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    4x in OgSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    4x in OgSelectionWidgetOptionsTest::testNonRequiredAudienceField from Drupal\Tests\og\Functional
    2x in EntityCreateAccessTest::testViewPermissionDoesNotGrantCreateAccess from Drupal\Tests\og\Kernel\Entity
    1x in OgSelectionTest::testSelectionHandler from Drupal\Tests\og\Kernel\Entity
    1x in OgSelectionTest::testSelectionHandlerResults from Drupal\Tests\og\Kernel\Entity
  22x: Providing settings under 'handler_settings' is deprecated in drupal:8.4.0 support for 'handler_settings' is removed from drupal:9.0.0. Move the settings in the root of the configuration array. See https://www.drupal.org/node/2870971
    10x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    4x in OgSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    4x in OgSelectionWidgetOptionsTest::testNonRequiredAudienceField from Drupal\Tests\og\Functional
    2x in EntityCreateAccessTest::testViewPermissionDoesNotGrantCreateAccess from Drupal\Tests\og\Kernel\Entity
    1x in OgSelectionTest::testSelectionHandler from Drupal\Tests\og\Kernel\Entity
    1x in OgSelectionTest::testSelectionHandlerResults from Drupal\Tests\og\Kernel\Entity
  3x: Drupal\Tests\BrowserTestBase::$defaultTheme is required in drupal:9.0.0 when using an install profile that does not set a default theme. See https://www.drupal.org/node/3083055, which includes recommendations on which theme to use.
    1x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    1x in OgSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
    1x in OgSelectionWidgetOptionsTest::testNonRequiredAudienceField from Drupal\Tests\og\Functional
  1x: EntityManagerInterface::getStorage() is deprecated in drupal:8.0.0 and is removed from drupal:9.0.0. Use \Drupal\Core\Entity\EntityTypeManager::getStorage() instead. See https://www.drupal.org/node/2549139
    1x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional
```

My problem with this PR is that it's ancient and I don't know any of the history. I've been trying to keep it in sync with develop (fix merge conflicts basically) for some time now but I have no idea why any of the hacks you mentioned are there (they all come from #181) so it's really hard to figure out where to go from here...

amitaibu commented:
My guess is that starting from scratch would be easier. The goal is to remove the custom field widget we've created in favor of the regular ER field.

The reason at the time for the OG complex field, was that we wanted site admins to more easily administer content - so they would see "My groups" (as select field), and "Other groups" (as auto-complete).

But it just adds more complexity to the code, where we could instead go with regular ER, and more complex widgets can be developed as contrib if they are really needed (which I'm not sure they are).

MPParsley committed:
Merge pull request #636 from Gizra/8.x-1.x

Merge develop into 570

MPParsley commented on og.module:
```diff
@@ -118,6 +118,7 @@ function og_entity_delete(EntityInterface $entity) {
  * Implements hook_entity_access().
  */
 function og_entity_access(EntityInterface $entity, $operation, AccountInterface $account) {
+
```
```suggestion
```

MPParsley committed:
undo remove whitespace

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php
#	tests/src/Functional/OgSelectionWidgetTest.php

MPParsley commented on tests/src/Functional/OgMultipleSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,192 @@
+      ->save();
+
+    // Create a role.
+    $role_storage = $this->container->get('entity.manager')->getStorage('user_role');
```
1x: EntityManagerInterface::getStorage() is deprecated in drupal:8.0.0 and is removed from drupal:9.0.0. Use \Drupal\Core\Entity\EntityTypeManager::getStorage() instead. See https://www.drupal.org/node/2549139
    1x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional

```suggestion
    $role_storage = $this->container->get('entity_type.manager')->getStorage('user_role');
```

MPParsley commented on tests/src/Functional/OgMultipleSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,192 @@
+   * {@inheritdoc}
+   */
+  public static $modules = ['node', 'og'];
+
```
3x: Drupal\Tests\BrowserTestBase::$defaultTheme is required in drupal:9.0.0 when using an install profile that does not set a default theme. See https://www.drupal.org/node/3083055, which includes recommendations on which theme to use.
    1x in OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from 

```suggestion

  /**
   * {@inheritdoc}
   */
  protected $defaultTheme = 'stark';

```

MPParsley commented on tests/src/Functional/OgSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,165 @@
+   * {@inheritdoc}
+   */
+  public static $modules = ['node', 'og'];
+
```
3x: Drupal\Tests\BrowserTestBase::$defaultTheme is required in drupal:9.0.0 when using an install profile that does not set a default theme. See https://www.drupal.org/node/3083055, which includes recommendations on which theme to use.
    1x in OgSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember from Drupal\Tests\og\Functional

```suggestion

  /**
   * {@inheritdoc}
   */
  protected $defaultTheme = 'stark';

```

MPParsley commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,181 @@
+   * {@inheritdoc}
+   */
+  public static $modules = ['node', 'og'];
+
```
3x: Drupal\Tests\BrowserTestBase::$defaultTheme is required in drupal:9.0.0 when using an install profile that does not set a default theme. See https://www.drupal.org/node/3083055, which includes recommendations on which theme to use.
    1x in OgSelectionWidgetOptionsTest::testNonRequiredAudienceField from Drupal\Tests\og\Functional

```suggestion

  /**
   * {@inheritdoc}
   */
  protected $defaultTheme = 'stark';

```

MPParsley committed:
Fix deprecation

MPParsley commented on tests/src/Kernel/EntityReference/OgSelectionConfigurationFormTest.php:
```diff
@@ -0,0 +1,96 @@
+
+    $form = \Drupal::formBuilder()->buildForm($form_object, $form_state);
+
+    $options = array_keys($form['settings']['handler']['handler_settings']['target_bundles']['#options']);
```
Providing settings under 'handler_settings' is deprecated in drupal:8.4.0 support for 'handler_settings' is removed from drupal:9.0.0. Move the settings in the root of the configuration array. See https://www.drupal.org/node/2870971
 
```suggestion
    $options = array_keys($form['settings']['handler']['target_bundles']['#options']);
```

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
+      return $query->condition($identifier_key, -1);
+    }
+
+    if ($this->currentUser->hasPermission('administer group')) {
```
#661

```suggestion
    if ($this->currentUser->hasPermission('administer organic groups')) {
```

MPParsley commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,186 @@
+    // Create users.
+    $this->groupMemberUser = $this->drupalCreateUser();
+    $this->groupOwnerUser = $this->drupalCreateUser();
+    $this->groupAdministratorUser = $this->drupalCreateUser(['administer group']);
```
#661

```suggestion
    $this->groupAdministratorUser = $this->drupalCreateUser(['administer organic groups']);
```

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -122,13 +133,32 @@ protected function setUp(): void {
+    ]);
+
+    $role
+      ->grantPermission('administer group')
```
#661

```suggestion
      ->grantPermission('administer organic groups')
```

MPParsley committed:
Renamed permission

#661 Rename the global permission to administer all groups to 'administer organic groups'

MPParsley commented on src/Og.php:
```diff
@@ -425,6 +431,8 @@ public static function getSelectionHandler(FieldDefinitionInterface $field_defin
    */
   public static function reset() {
     static::$cache = [];
+
+    \Drupal::service('og.access')->reset();
```
OgAccessInterface::reset() is deprecated in og:8.1.0-alpha6 and is removed from og:8.1.0-beta1. The static cache has been removed and this no longer server any purpose. See https://github.com/Gizra/og/issues/654

```suggestion
```

MPParsley committed:
Remove deprecated method reset()

MPParsley commented on src/Og.php:
```diff
@@ -405,17 +403,25 @@ public static function getSelectionHandler(FieldDefinitionInterface $field_defin
-
-    // Override with passed $options.
-    $options = NestedArray::mergeDeep($default_options, $options);
+      'handler_settings' => $field_definition->getSetting('handler_settings'),
+      'entity' => $entity,
+    ];
```
Providing settings under 'handler_settings' is deprecated in drupal:8.4.0 support for 'handler_settings' is removed from drupal:9.0.0. Move the settings in the root of the configuration array. See https://www.drupal.org/node/2870971 

```suggestion
      'entity' => $entity,
    ] + $field_definition->getSetting('handler_settings');
```

MPParsley committed:
Deprecate handler_settings

Providing settings under 'handler_settings' is deprecated in drupal:8.4.0 support for 'handler_settings' is removed from drupal:9.0.0. Move the settings in the root of the configuration array. See https://www.drupal.org/node/2870971

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -122,13 +133,32 @@ protected function setUp(): void {
     ];
     $this->selectionPluginManager->getInstance($options);
     $this->selectionHandler = $this->selectionPluginManager->getSelectionHandler($this->fieldDefinition);
+    // @todo: remove deprecated call to Og::getSelectionHandler.
+    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```
Remove deprecated call to Og::getSelectionHandler.

```suggestion
```

MPParsley committed:
Remove deprecated call to Og::getSelectionHandler

MPParsley commented on tests/src/Kernel/EntityReference/OgSelectionConfigurationFormTest.php:
```diff
@@ -0,0 +1,96 @@
+
+    $form = \Drupal::formBuilder()->buildForm($form_object, $form_state);
+
+    $options = array_keys($form['settings']['handler']['target_bundles']['#options']);
```
```suggestion
    $options = array_keys($form['settings']['handler']['handler_settings']['target_bundles']['#options']);
```

MPParsley committed:
Fix undefined index target_bundles

MPParsley commented on og.module:
```diff
@@ -201,23 +221,40 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
-
-    if ($handler->getInstance($options)) {
-      return AccessResult::neutral();
+    $handler = $handler->getInstance($options);
+
+    // @todo: remove deprecated call to Og::getSelectionHandler.
+    $handler = Og::getSelectionHandler($field_definition);
```
Remove deprecated call to Og::getSelectionHandler

```suggestion
    $handler = $handler->getInstance($options);
```

MPParsley committed:
Remove deprecated call to Og::getSelectionHandler

MPParsley commented on og.module:
```diff
@@ -201,23 +221,40 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
-
-    if ($handler->getInstance($options)) {
-      return AccessResult::neutral();
+    $handler = $handler->getInstance($options);
```
```suggestion
    $handler = $handler->getInstance($options);

    // @todo: remove deprecated call to Og::getSelectionHandler.
    $handler = Og::getSelectionHandler($field_definition);
```

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -123,12 +134,29 @@ protected function setUp(): void {
     $this->selectionPluginManager->getInstance($options);
     $this->selectionHandler = $this->selectionPluginManager->getSelectionHandler($this->fieldDefinition);
```
```suggestion
    $this->selectionHandler = $this->selectionPluginManager->getSelectionHandler($this->fieldDefinition);

    // @todo: remove deprecated call to Og::getSelectionHandler.
    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```

MPParsley committed:
Revert Og::getSelectionHandler

damienmckenna commented on og.module:
```diff
@@ -248,6 +248,9 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
     $handler = \Drupal::service('plugin.manager.entity_reference_selection');
     $handler = $handler->getInstance($options);
 
+    // @todo: remove deprecated call to Og::getSelectionHandler.
```
FYI `@todo` docblock comments shouldn't have a colon and should start with a capital letter.
```suggestion
    // @todo Remove deprecated call to Og::getSelectionHandler.
```

MPParsley replied:
Hi @damienmckenna, thanks for the input. I tried removing the call to Og::getSelectionHandler but it seems like this PR has modified its behavior and it's still needed.

PS: did you know you can provide code suggestions on Github with the ± icon in the editor?

damienmckenna commented on og.module:
```diff
@@ -156,6 +156,21 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
+  if (count($audience_fields) === 1) {
+    $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity_type_id, $bundle_id);
+    foreach ($field_definitions as $field_definition) {
+
```
The extra blank line isn't needed.
```suggestion
```

damienmckenna commented on og.module:
```diff
@@ -300,9 +347,24 @@ function og_field_widget_info_alter(array &$info) {
  * be rebuilt via RouteBuilder::setRebuildNeeded.
  */
 function og_entity_type_alter(array &$entity_types) {
+
```
The blank line isn't necessary.
```suggestion
```

damienmckenna commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
 
-    $identifier_key = $definition->getKey('id');
+    if (empty($this->configuration['entity'])) {
+      // @todo: Find out why we have this scenario.
```
Coding standards.
```suggestion
      // @todo Find out why we have this scenario.
```

damienmckenna commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -128,4 +210,29 @@ protected function getUserGroups() {
+    $bundles_info = $this->entityTypeBundleInfo->getBundleInfo($entity_type_id);
+
+    if ($entity_type->hasKey('bundle')) {
+
```
The blank line is unnecessary.
```suggestion
```

damienmckenna commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,186 @@
+   */
+  protected $group2;
+
+
```
One blank line too many.
```suggestion
```

damienmckenna commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,186 @@
+    $this->drupalGet('node/add/group_content');
+    // The user can't post group content site wide so the audience is required
+    // field.
+    // todo: is that correct?
```
Minor coding standards adjustment.
```suggestion
    // todo Is that correct?
```

MPParsley committed:
Apply suggestions from code review

Co-authored-by: Damien McKenna <damien@mc-kenna.com>

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +20,34 @@ public function validate($value, Constraint $constraint) {
+    $entity = $this->context->getRoot()->getValue();
+
+    /** @var \Drupal\Core\Access\AccessResult $access */
+    // @todo: Refactor the permission format in #510.
+    $permission = "create {$entity->bundle()} content";
+    $user = \Drupal::currentUser()->getAccount();
+    $access = \Drupal::service('og.access')->userAccessEntity($permission, $group, $user);
```
Split up OgAccess::userAccessEntity() #673 

```suggestion
    $user = \Drupal::currentUser()->getAccount();
    $access = \Drupal::service('og.access')-> userAccessGroupContentEntityOperation('create', $entity, $group, $user);
```

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Split up OgAccess::userAccessEntity

See #673.

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +20,32 @@ public function validate($value, Constraint $constraint) {
+
+    /** @var \Drupal\Core\Access\AccessResult $access */
+    $user = \Drupal::currentUser()->getAccount();
+    $access = \Drupal::service('og.access')-> userAccessGroupContentEntityOperation('create', $entity, $group, $user);
```
```suggestion
    $access = \Drupal::service('og.access')-> userAccessGroupContentEntityOperation('create', $group, $entity, $user);
```

MPParsley committed:
Fix order of parameters

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -79,37 +153,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
-      // Don't include the groups, the user doesn't have create permission.
-      foreach ($user_groups as $group) {
+    foreach ($this->getUserGroups() as $group) {
+      // Check user has "create" permission on this entity.
+      $permission = "create $bundle content";
+      if ($this->ogAccess->userAccess($group, $permission, $this->currentUser)->isAllowed()) {
```
```suggestion
      // Check if the current user can perform the "create" operation on this entity.
      if ($this->ogAccess->userAccessGroupContentEntityOperation('create', $group, $entity, $this->currentUser)) {
```

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +20,32 @@ public function validate($value, Constraint $constraint) {
+
+    /** @var \Drupal\Core\Access\AccessResult $access */
+    $user = \Drupal::currentUser()->getAccount();
+    $access = \Drupal::service('og.access')-> userAccessGroupContentEntityOperation('create', $group, $entity, $user);
```
```suggestion
    $access = \Drupal::service('og.access')->userAccessGroupContentEntityOperation('create', $group, $entity, $user);
```

MPParsley committed:
Fix syntax error

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraint.php:
```diff
@@ -23,4 +23,11 @@ class ValidOgMembershipReferenceConstraint extends Constraint {
+   *
+   * @var string
+   */
+  public $notAllowedToPostInGroup = 'You are not allowed to post content in the group %label';
```
```suggestion
  public $notAllowedToPostInGroup = 'You are not allowed to post content in the group %label.';
```

MPParsley committed:
Added period.

MPParsley committed:
Inject services

- Drupal\Core\Session\AccountInterface
- Drupal\og\OgAccessInterface

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraintValidator.php:
```diff
@@ -0,0 +1,110 @@
+   */
+  public function validate($entity, Constraint $constraint) {
+    /* @var \Drupal\Core\Entity\ContentEntityBase $entity */
+    if (!Og::isGroupContent($entity->getEntityTypeId(), $entity->bundle())) {
```
Og::isGroupContent:
```
  public static function isGroupContent($entity_type_id, $bundle_id) {
    return \Drupal::service('og.group_audience_helper')->hasGroupAudienceField($entity_type_id, $bundle_id);
  }
```

```suggestion
    if (!$this->groupAudienceHelper->hasGroupAudienceField($entity->getEntityTypeId(), $entity->bundle())) {
```

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +60,32 @@ public function validate($value, Constraint $constraint) {
+
+    /** @var \Drupal\Core\Access\AccessResult $access */
+    $user = $this->currentUser->getAccount();
+    $access = $this->ogAccess->userAccessGroupContentEntityOperation('create', $group, $entity, $user);
```
Revert injection of OgAccess service.

```suggestion
    $access = \Drupal::service('og.access')->userAccessGroupContentEntityOperation('create', $group, $entity, $user);
```

MPParsley committed:
Revert injection of OgAccess

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -20,21 +60,32 @@ public function validate($value, Constraint $constraint) {
+    $entity = $this->context->getRoot()->getValue();
+
+    /** @var \Drupal\Core\Access\AccessResult $access */
+    $user = $this->currentUser->getAccount();
```
Revert injection of current user
```suggestion
    $user = \Drupal::currentUser()->getAccount();
```

MPParsley committed:
Revert injection of current user

MPParsley commented on tests/src/Functional/OgSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,170 @@
+    $this->drupalGet('node/add/group_content');
+    $this->submitForm($edit, 'Save');
+
+    // When using 8.4, ValidReferenceConstraintValidator is prevent from
+    // ValidOgMembershipReferenceConstraintValidator message to appear. We need
+    // to see how we can override that so the user would have a better
+    // understanding why the reference is invalid.
+    try {
+      $this->assertSession()->pageTextContains('You are not allowed to post content in the group ' . $this->group2->label());
+    }
+    catch (ResponseTextException $e) {
+      $this->assertSession()->pageTextContains('This entity (node: ' . $this->group2->id() . ') cannot be referenced.');
+    }
+    catch (\Exception $e) {
+      throw new \Exception('Both of the errors for the invalid group reference did not appear on the screen.');
+    }
```
Remove hack for Drupal 8.4
```suggestion
    $this->assertSession()->pageTextContains('You are not allowed to post content in the group ' . $this->group2->label() . '.');
```

MPParsley committed:
Remove hack for Drupal 8.4

MPParsley committed:
Revert dependency injections

MPParsley commented on tests/src/Functional/OgSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,158 @@
+    $this->drupalGet('node/add/group_content');
+    $this->submitForm($edit, 'Save');
+
+    $this->assertSession()->pageTextContains('You are not allowed to post content in the group ' . $this->group2->label() . '.');
```
Revert hack
```suggestion
    // When using 8.4, ValidReferenceConstraintValidator is prevent from
    // ValidOgMembershipReferenceConstraintValidator message to appear. We need
    // to see how we can override that so the user would have a better
    // understanding why the reference is invalid.
    try {
      $this->assertSession()->pageTextContains('You are not allowed to post content in the group ' . $this->group2->label());
    }
    catch (ResponseTextException $e) {
      $this->assertSession()->pageTextContains('This entity (node: ' . $this->group2->id() . ') cannot be referenced.');
    }
    catch (\Exception $e) {
      throw new \Exception('Both of the errors for the invalid group reference did not appear on the screen.');
    }
```

MPParsley committed:
Revert hack for ValidReferenceConstraintValidator

Commit:
Inject services.

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipReferenceConstraintValidator.php:
```diff
@@ -2,14 +2,78 @@
  * Checks if referenced entities are valid.
  */
-class ValidOgMembershipReferenceConstraintValidator extends ConstraintValidator {
+class ValidOgMembershipReferenceConstraintValidator extends ConstraintValidator implements ContainerInjectionInterface {
```
Tests are failing once this is added: "implements ContainerInjectionInterface"

Commit:
Fix tests

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -123,12 +134,32 @@ protected function setUp(): void {
-    $this->user1 = User::create(['name' => $this->randomString()]);
-    $this->user1->save();
+    // @todo: remove deprecated call to Og::getSelectionHandler.
+    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```
```suggestion
//    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```

MPParsley commented on og.module:
```diff
@@ -201,23 +220,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
+    $handler = $handler->getInstance($options);
+
+    // @todo Remove deprecated call to Og::getSelectionHandler.
+    $handler = Og::getSelectionHandler($field_definition);
```
```suggestion
//    $handler = Og::getSelectionHandler($field_definition);
```

MPParsley committed:
Test deprecation of Og:getSelectionHandler

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -70,7 +144,7 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
 
     if ($bundle_key = $definition->getKey('bundle')) {
-      $bundles = Og::groupTypeManager()->getAllGroupBundles($target_type);
+      $bundles = Og::groupTypeManager()->getGroupBundlesByEntityType($target_type);
```
```suggestion
      $bundles = Og::groupTypeManager()->getGroupBundleIdsByEntityType($target_type);
```

MPParsley committed:
Use existing method getGroupBundleIdsByEntityType

MPParsley committed:
Revert getAllGroupBundles

MPParsley committed:
Remove getGroupBundlesByEntityType

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php

MPParsley commented on og.module:
```diff
@@ -203,23 +222,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
+    $handler = $handler->getInstance($options);
+
+    // @todo Remove deprecated call to Og::getSelectionHandler.
+//    $handler = Og::getSelectionHandler($field_definition);
```
```suggestion
    $handler = Og::getSelectionHandler($field_definition);
```

MPParsley committed:
Revert deprecation

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -125,12 +136,32 @@ protected function setUp(): void {
-    $this->user1 = User::create(['name' => $this->randomString()]);
-    $this->user1->save();
+    // @todo: remove deprecated call to Og::getSelectionHandler.
+//    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```
```suggestion
    $this->selectionHandler = Og::getSelectionHandler($this->fieldDefinition);
```

MPParsley committed:
Revert deprecation

MPParsley committed:
Fix strict types

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x'

Resolve merge conflicts.

MPParsley commented on tests/src/Kernel/Entity/OgSelectionTest.php:
```diff
@@ -125,12 +136,32 @@ protected function setUp(): void {
-    // Create two users.
-    $this->user1 = User::create(['name' => $this->randomString()]);
-    $this->user1->save();
+    // @todo: remove deprecated call to Og::getSelectionHandler.
```
```suggestion
    // @todo Remove deprecated call to Og::getSelectionHandler.
```

MPParsley commented on tests/src/Functional/OgSelectionWidgetOptionsTest.php:
```diff
@@ -0,0 +1,187 @@
+    $this->drupalGet('node/add/group_content');
+    // The user can't post group content site wide so the audience is required
+    // field.
+    // todo Is that correct?
```
```suggestion
    // @todo Is that correct?
```

MPParsley commented on src/Plugin/OgFields/AudienceField.php:
```diff
@@ -60,14 +60,9 @@ public function getFieldBaseDefinition(array $values = []) {
    * {@inheritdoc}
    */
   public function getFormDisplayDefinition(array $values = []) {
+    // @todo: isn't it taken from the definition in OgStandardReferenceItem?
```
```suggestion
    // @todo Isn't it taken from the definition in OgStandardReferenceItem?
```

MPParsley commented on src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraintValidator.php:
```diff
@@ -0,0 +1,112 @@
+   * {@inheritdoc}
+   */
+  public function validate($entity, Constraint $constraint) {
+    /* @var \Drupal\Core\Entity\ContentEntityBase $entity */
```
```suggestion
    /** @var \Drupal\Core\Entity\ContentEntityBase $entity */
```

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -81,37 +155,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
+    $identifier_key = $definition->getKey('id');
+
+    if ($this->currentUser->isAnonymous()) {
+      // @todo: Check if anonymous users should have access to any referenced
```
```suggestion
      // @todo Check if anonymous users should have access to any referenced
```

MPParsley committed:
Apply suggestions from code review

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget
